modules/daemon/daemon.js

```diff
@@ -0,0 +1,135 @@
+const electron = require('electron');
+const log      = require('electron-log');
+const spawn    = require('child_process').spawn;
+const rxIpc    = require('rx-ipc-electron/lib/main').default;
+
+const _options      = require('../options');
+const rpc           = require('../rpc/rpc');
+const cookie        = require('../rpc/cookie');
+const daemonManager = require('../daemon/daemonManager');
+
+let daemon;
+let exitCode = 0;
+
+function daemonData(data, logger) {
+  logger(data.toString().replace(/[\n\r]/g, ""));
+}
+
+exports.start = function(wallets, callback) {
+  return (new Promise((resolve, reject) => {
+
+    let   options    = _options.get();
+    const daemonPath = options.customdaemon
+                     ? options.customdaemon
+                     : daemonManager.getPath();
+
+    exports.check().then(() => {
+      log.info('daemon already started');
+      resolve(undefined);
+
+    }).catch(() => {
+
+      wallets = wallets.map(wallet => `-wallet=${wallet}`);
+      log.info(`starting daemon ${daemonPath} ${process.argv} ${wallets}`);
+
+      const child = spawn(daemonPath, [...process.argv, ...wallets])
+      .on('close', code => {
+        if (code !== 0) {
+          reject();
+          log.error(`daemon exited with code ${code}.\n${daemonPath}\n${process.argv}`);
+        } else {
+          log.info('daemon exited successfully');
+        }
+        electron.app.quit();
+      })
+
+      child.stdout.on('data', data => daemonData(data, console.log));
+      child.stderr.on('data', data => daemonData(data, console.log));
+
+      daemon = child;
+      exports.wait(wallets, callback).then(() => resolve());
+    });
+
+  }));
+}
+
+exports.wait = function(wallets, callback) {
+  return new Promise((resolve, reject) => {
+
+    const maxRetries  = 10; // Some slow computers...
+    let   retries     = 0;
+    let   errorString = '';
+
+    const daemonStartup = () => {
+      exports.check()
+      .then(() => { callback(); resolve(); })
+      .catch(() => {
+        if (exitCode === 0 && retries < maxRetries) {
+          setTimeout(daemonStartup, 1000);
+        }
+      });
+
+      retries++;
+      if (exitCode || retries >= maxRetries) {
```

Comment:
`retries >= maxRetries`
This is also quite weird

Reply:
I agree, but I think it works. I'll test it and maybe see why tomorrow

Reply:
it actually makes sense. there's 10 retries, one every second, and when we exceed that we check for `-reindex` need, if not we exit. I think `reindex` could be done each time to boot faster, but I'm also not too reassured by the fact that there's 2 recursive calls in this function that could fire in the same call so i've protected the first one with a return.